File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;
class Solution {
    class State {
        int idx, val;
        
        State(int idx, int val) {
            this.idx = idx;
            this.val = val;
        }
    }
    
    public int solution(int[] numbers, int target) {
        return bfs(numbers, target);
    }
    
    int bfs(int[] numbers, int target) {
        int answer = 0;
        
        // init
        Queue<State> q = new LinkedList<>();
         
        q.offer(new State(0, numbers[0]));
        q.offer(new State(0, -numbers[0]));
        
        while (!q.isEmpty()) {
            State current = q.poll();
            
            if (current.idx == numbers.length - 1) {
                if (current.val == target) {
                    answer++;
                }
                continue;
            }
            
            int nextIdx = current.idx + 1;
            
            q.offer(new State(nextIdx, current.val + numbers[nextIdx]));
            q.offer(new State(nextIdx, current.val - numbers[nextIdx]));
        }
    
        return answer;
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
answer = 0

def solution(numbers, target):
    
    # 갈 수 있는 루트 : 더하거나 / 빼기 
    # 목표 : 가능한 모든 경우의 수 
    # 가능한 경우의 수는 연산자를 + 하거나 - 하는거라서 brute force 가 가능 -> dfs
    
    dfs(0, 0, numbers, target)
    return answer

# idx: 현재 보고 있는 숫자 인덱스 
def dfs(idx, currentSum, numbers, target):
    global answer
    
    # base case 
    if idx == len(numbers):
        if currentSum == target:
            answer += 1
        return 

    # 이 숫자를 더할지 뺄지 
    dfs(idx + 1, currentSum + numbers[idx], numbers, target)
    dfs(idx + 1, currentSum - numbers[idx], numbers, target)